[libc++][doc] Use installed std modules in external projects.

a858438680 opened the pull request:
The original libc++ document about using std modules in external projects is outdated. Loot at #80231 for more details.

This pull request update the document. The new document tells how to use the current installed std modules without toolchains' support.

a858438680 committed:
[libc++][doc] Use installed std modules in external projects.

a858438680 committed:
[libc++][doc] Use installed std modules in external projects.

ChuanqiXu9 commented on libcxx/docs/Modules.rst:
```diff
@@ -105,36 +105,36 @@ Users need to be able to build their own BMI files.
-Currently this requires a local build of libc++ with modules enabled. Since
-modules are not part of the installation yet, they are used from the build
-directory. First libc++ needs to be build with module support enabled.
+Currently this requires a local build of libc++ with modules installation enabled.
+Since modules are not installed by default. You can build and install the modules
+to ``<install_prefix>`` with the following commands.
```
```suggestion
This requires the system you're distributes the installed std modules. Otherwise...
```

a858438680 replied:
Thanks for your review. But I do not understand why you want me to delete these three lines. Could you please explain more in detail?
And also, The line you suggest me to add is not complete and seems to have a grammer error. Could you please tell me what you want to express?

ChuanqiXu9 replied:
Sorry. I mean it may be better to not say "the modules are not installed by default". Since there may be system vendors enable it.

a858438680 replied:
Ok, got it. So you mean "requires the system you're using distributes ..." right? And "Otherwise, you need to build and install ..." right? I think this suggestion is fine. But I also want to hear from main author of the original page @mordante.

mordante replied:
We shouldn't change this in this fashion. The current described way for building is broken, this should be fixed. So that way remains a viable option. The goals of this patch is to add a second option based on installed modules.

poyaoc97 commented on libcxx/docs/Modules.rst:
```diff
@@ -168,58 +168,95 @@ This is a small sample program that uses the module ``std``. It consists of a
-  # For example from, the headers <cassert> and <version>.
-  add_compile_options($<$<COMPILE_LANGUAGE:CXX>:-isystem>)
-  add_compile_options($<$<COMPILE_LANGUAGE:CXX>:${LIBCXX_BUILD}/include/c++/v1>)
+  add_library(std)
```
Missing "STATIC"?

a858438680 replied:
I will add it

mordante commented on libcxx/docs/Modules.rst:
```diff
@@ -105,36 +105,36 @@ Users need to be able to build their own BMI files.
 .. code-block:: bash
 
-  $ git clone https://github.com/llvm/llvm-project.git
+  $ git clone https://github.com/llvm/llvm-project.git --depth 1
```
I prefer to keep the instructions generic, if users like `--depth 1` they are free to do so.

a858438680 replied:
OK

mordante commented on libcxx/docs/Modules.rst:
```diff
@@ -105,36 +105,36 @@ Users need to be able to build their own BMI files.
 
   int main() {
-    std::cout << "Hello modular world\n";
+    std::println("Hello modular world");
```
please don't make unrelated changes in the same patch.

a858438680 replied:
OK, will change it back

mordante reviewed (changes requested):
Thanks for the patch. I really like to keep the existing method and add a new method. I'll make a patch to fix to reintroduce the accidentally removed CMakeLists.txt.

mordante committed:
[libc++][modules] Re-add build dir CMakeLists.txt. (#81370)

This CMakeLists.txt is used to build modules without build system
support. This was removed in d06ae33ec32122bb526fb35025c1f0cf979f1090.
This is used in the documentation how to use modules.

Made some minor changes to make it work with the std.compat module using
the std module.

Note the CMakeLists.txt in the build dir should be removed once build
system support is generally available.

Commit:
[flang] fix one more compilation issue on aarch64. (#83258)

I used `class` instead of `struct` for `SymbolMap`.

labrinea committed:
[clang] Remove unused lambda capture. (#83550)

Fixes the `sanitizer-x86_64-linux-android` buildbot.

mordante commented:
@a858438680 are you still working on this patch?

a858438680 commented:
> @a858438680 are you still working on this patch?

Yes, but I missed your previous comments. I will make some changes according to your comments soon.

a858438680 committed:
[libc++][doc] Use installed std modules with the previous version kept.

a858438680 commented:
@mordante Hi, I have changed this commit according the your comments.

mordante commented:
> @mordante Hi, I have changed this commit according the your comments.

Thanks! Can you rebase the patch, there are merge conflicts. I prefer to review without these conflicts.

a858438680 committed:
[libc++][doc] Fix libc++ build script

a858438680 committed:
[libc++][doc] Remove the redundant note.

a858438680 commented:
@mordante Hi, there is no conflict now.

mordante commented on libcxx/docs/Modules.rst:
```diff
@@ -107,34 +107,40 @@ Users need to be able to build their own BMI files.
-  $ ninja -C build
+  $ cmake -G Ninja -S runtimes -B build -DLIBCXX_INSTALL_MODULES=ON -DLLVM_ENABLE_RUNTIMES="libcxx;libcxxabi;libunwind"
+  $ cmake --build build -- -j $(nproc)
+  $ cmake --install build --prefix <install_prefix>
```
I really prefer to keep the `ninja -C build` and add `ninja install -C build`. This is the style we use in other parts of the libc++ documentation.

mordante commented on libcxx/docs/Modules.rst:
```diff
@@ -168,51 +174,82 @@ This is a small sample program that uses the module ``std``. It consists of a
-    std
-    URL "file://${LIBCXX_BUILD}/modules/c++/v1/"
+    std_module
+    URL "file://${LIBCXX_INSTALLED_DIR}/share/libc++/v1"
```
This part of the example should use the `${LIBCXX_BUILD}` right? Are the other parts of the changes in this part intended?

a858438680 committed:
[libc++][doc] Recover unintended changes of original docs.

mordante commented on libcxx/docs/Modules.rst:
```diff
@@ -225,6 +231,137 @@ Building this project is done with the following steps, assuming the files
           building the ``std`` target using multiple jobs.
 
+This is another small sample program that uses the module ``std`` from
+installation directory. It consists of a ``CMakeLists.txt``, an
+``std.cmake``, and a ``main.cpp`` file. The ``main.cpp`` is the same as
+the previous example.
```
```suggestion
the installation directory. It consists of a ``CMakeLists.txt``,
a ``std.cmake``, and the ``main.cpp`` from the previous example.
```

mordante commented on libcxx/docs/Modules.rst:
```diff
@@ -225,6 +231,137 @@ Building this project is done with the following steps, assuming the files
+  $ ninja -C build
+  $ build/main
+
+.. warning:: You need more than clang itself to build a project using modules.
+             Specifically, you will need ``clang-scan-deps``. For example, in Ubuntu, you
+             need to use ``sudo ./llvm.sh 17 all`` rather than ``sudo ./llvm.sh 17`` showed
+             in `LLVM Debian/Ubuntu nightly packages <https://apt.llvm.org>`__ to install
+             essential components to build this project.
+
```
I would remove this part. This is also required when using the build directory.

mordante reviewed (changes requested):
Thanks I finally had time to have a good look. I tested in locally and the instructions work nicely :-)

There have been some other changes in this file, can you rebase before updating this review?

mordante commented:
CMake currently has experimental support for modules in their master branch. I've update our modules status page. Do you still want to pursue this patch? If not I'll close the patch.

a858438680 commented:
> CMake currently has experimental support for modules in their master branch. I've update our modules status page. Do you still want to pursue this patch? If not I'll close the patch.

Actually I looked the newest document. And I think this patch is not maintainable, giving that cmake now has their official method. So I don't want to pursue this patch. It's completely ok for me to close this.

mordante commented:
> > CMake currently has experimental support for modules in their master branch. I've update our modules status page. Do you still want to pursue this patch? If not I'll close the patch.
> 
> Actually I looked the newest document. And I think this patch is not maintainable, giving that cmake now has their official method. So I don't want to pursue this patch. It's completely ok for me to close this.

Thanks then I'll close this.
 
> And Actually I tried to rebase this path to the main branch. But now I don't know how to do it.

You need to resolve the conflicts to be able to rebase. If you want to know more about this, let me know.
